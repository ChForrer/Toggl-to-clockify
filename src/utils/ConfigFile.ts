import * as fs from 'fs';
import * as path from 'path';
import JsonFile from './JsonFile';
import { Config } from '../types/common';

export const defaultSettings = {
  email: '',
  togglApiToken: '',
  clockifyApiToken: '',
  workspaces: [],
};

export default class ConfigFile {
  public static validateFilePath(filePath?: string | null): string | null {
    const configFilePath = path.resolve(process.cwd(), 't2c.json');
    if (filePath && fs.existsSync(filePath)) return filePath;
    if (fs.existsSync(configFilePath)) return configFilePath;
    return null;
  }

  public static loadEntriesFromFile(filePath: string): Config {
    const jsonFile = new JsonFile(filePath);
    const configFileContents = jsonFile.readSync();
    if (configFileContents === null) return defaultSettings;
    return configFileContents;
  }

  public static async generateFile(targetPath: string) {
    const configFile = new JsonFile(targetPath);
    const generatedContents = {
      ...defaultSettings,
      workspaces: [
        {
          name: '',
          years: [],
        },
      ],
    };
    await configFile.write(generatedContents);
  }
}
