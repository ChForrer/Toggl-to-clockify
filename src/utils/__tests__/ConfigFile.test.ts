import * as fs from 'fs';
import * as path from 'path';
import * as jsonFile from 'jsonfile';
import mock from 'mock-fs';
import ConfigFile from '../ConfigFile';

describe('The ConfigFile class', () => {
  describe('the validateFilePath() method', () => {
    beforeAll(() => {
      mock({
        't2c.json': mock.file({
          content: '{}',
        }),
      });
    });

    afterAll(() => {
      mock.restore();
    });

    test(`returns the default file path if it exists`, () => {
      const result = ConfigFile.validateFilePath() as string;
      expect(path.dirname(result)).toEqual(process.cwd());
      expect(path.basename(result)).toEqual('t2c.json');
    });
  });

  describe('the loadEntriesFromFile() method', () => {
    test(`matches its snapshot when the config file is valid`, () => {
      const testConfigPath = path.resolve(
        __dirname,
        '..',
        '__fixtures__',
        't2c.json',
      );
      const result = ConfigFile.loadEntriesFromFile(testConfigPath);
      expect(result).toMatchSnapshot();
    });

    test(`matches its snapshot when the config file doesn't exist`, () => {
      const result = ConfigFile.loadEntriesFromFile('');
      expect(result).toMatchSnapshot();
    });
  });

  describe('the generateFile() method', () => {
    const testFilePath = `${__dirname}/generateFile.json`;

    test(`file contents matches its snapshot`, async () => {
      expect.assertions(1);
      await ConfigFile.generateFile(testFilePath);
      const contents = jsonFile.readFileSync(testFilePath);
      fs.unlinkSync(testFilePath);
      expect(contents).toMatchSnapshot();
    });
  });
});
