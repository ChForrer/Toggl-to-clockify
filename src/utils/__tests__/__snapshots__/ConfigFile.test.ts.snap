// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ConfigFile class the generateFile() method file contents matches its snapshot 1`] = `
Object {
  "clockifyApiToken": "",
  "email": "",
  "togglApiToken": "",
  "workspaces": Array [
    Object {
      "name": "",
      "years": Array [],
    },
  ],
}
`;

exports[`The ConfigFile class the loadEntriesFromFile() method matches its snapshot when the config file doesn't exist 1`] = `
Object {
  "clockifyApiToken": "",
  "email": "",
  "togglApiToken": "",
  "workspaces": Array [],
}
`;

exports[`The ConfigFile class the loadEntriesFromFile() method matches its snapshot when the config file is valid 1`] = `
Object {
  "clockifyApiToken": "clockify123abc",
  "email": "testing@test.com",
  "togglApiToken": "toggl123abc",
  "workspaces": Array [
    Object {
      "name": "Test WS 1",
      "years": Array [
        2017,
        2018,
      ],
    },
    Object {
      "name": "Test WS 2",
      "years": Array [
        2017,
        2018,
      ],
    },
  ],
}
`;
